flang/lib/Optimizer/CodeGen/Target.cpp

```diff
@@ -0,0 +1,189 @@
+//===-- Target.cpp --------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Coding style: https://mlir.llvm.org/getting_started/DeveloperGuide/
+//
+//===----------------------------------------------------------------------===//
+
+#include "Target.h"
+#include "flang/Optimizer/Dialect/FIRType.h"
+#include "flang/Optimizer/Support/KindMapping.h"
+#include "mlir/IR/StandardTypes.h"
+#include "mlir/IR/TypeRange.h"
+#include "llvm/ADT/Triple.h"
+
+#define DEBUG_TYPE "flang-codegen-target"
+
+using namespace fir;
+
+// Reduce a REAL/float type to the floating point semantics.
+static const llvm::fltSemantics &floatToSemantics(KindMapping &kindMap,
+                                                  mlir::Type type) {
+  assert(isa_real(type));
+  if (auto ty = type.dyn_cast<fir::RealType>())
+    return kindMap.getFloatSemantics(ty.getFKind());
+  return type.cast<mlir::FloatType>().getFloatSemantics();
+}
+
+namespace {
+template <typename S>
+struct GenericTarget : public CodeGenSpecifics {
+  using CodeGenSpecifics::CodeGenSpecifics;
+  using AT = CodeGenSpecifics::Attributes;
+
+  mlir::Type complexMemoryType(mlir::Type eleTy) const override {
+    assert(fir::isa_real(eleTy));
+    // { t, t }   struct of 2 eleTy
+    mlir::TypeRange range = {eleTy, eleTy};
+    return mlir::TupleType::get(range, eleTy.getContext());
+  }
+
+  mlir::Type boxcharMemoryType(mlir::Type eleTy) const override {
+    auto idxTy = mlir::IntegerType::get(S::defaultWidth, eleTy.getContext());
+    auto ptrTy = fir::ReferenceType::get(eleTy);
+    // { t*, index }
+    mlir::TypeRange range = {ptrTy, idxTy};
+    return mlir::TupleType::get(range, eleTy.getContext());
+  }
+
+  Marshalling boxcharArgumentType(mlir::Type eleTy, bool sret) const override {
+    CodeGenSpecifics::Marshalling marshal;
+    auto idxTy = mlir::IntegerType::get(S::defaultWidth, eleTy.getContext());
+    auto ptrTy = fir::ReferenceType::get(eleTy);
+    marshal.emplace_back(ptrTy, AT{});
+    // Return value arguments are grouped as a pair. Others are passed in a
+    // split format with all pointers first (in the declared position) and all
+    // LEN arguments appended after all of the dummy arguments.
+    // NB: Other conventions/ABIs can/should be supported via options.
+    marshal.emplace_back(idxTy, AT{0, {}, {}, /*append=*/!sret});
```

Comment:
So far the returned charbox were already passed as split argument in lowering, that make sens to update this to use what you are doing here. However, character returns are caller allocated, I saw in your change that there is a `FixupTy::Codes::ReturnAsStore` that seems to allocate the return on the callee side, is it possible to insert allocation on caller side for character returns or should lowering still take care of this ?

Reply:
A Fortran function that returns a CHARACTER value is intended (modulo my bugs) to be passed as a SRET argument of Boxchar type. That means the caller has to pass (and allocate) the space and pass the LEN value. This differs from a non-SRET Boxchar argument in how the argument is ordered in the signature only. The called function should play no role in allocating space for the result (or arguments).